src/db.rs

```diff
@@ -0,0 +1,590 @@
+            update_stats_stmt.execute(&[block_size])?;
+            stats.count -= 1;
+            stats.size -= block_size as u64;
+        }
+        delete_stmt.execute(&[id])?;
+        n += 1;
+    }
+    cache_tracker.delete_ids(&ids[0..n]);
+    Ok(n == ids.len() || !size_targets.exceeded(&stats))
+}
+
+/// deletes the orphaned blocks.
+///
+/// orphaned blocks are blocks from the blocks table that do not have a corresponding id in the
+/// cid table and in the other metadata table. They are unreachable.
+///
+/// The reason for deleting them incrementally is that deleting blocks can be an expensive operation
+/// in sqlite, and we want to minimize gc related interruptions.
+///
+/// note that the execution time limit is not entirely accurate, because in many cases the cost of
+/// deleting blocks will only be fully felt when doing the commit of the transaction.
+pub(crate) fn incremental_delete_orphaned(
+    txn: &Transaction,
+    min_blocks: usize,
+    max_duration: Duration,
+) -> rusqlite::Result<bool> {
+    let t0 = Instant::now();
+    let ids: Vec<i64> = log_execution_time("determine_orphaned", Duration::from_secs(1), || {
+        txn.prepare_cached(
+            "SELECT block_id FROM blocks WHERE block_id NOT IN (SELECT id FROM cids)",
+        )?
+        .query_map(NO_PARAMS, |row| row.get(0))?
+        .collect::<rusqlite::Result<_>>()
+    })?;
+    let mut delete_stmt = txn.prepare_cached("DELETE FROM blocks WHERE block_id = ?")?;
+    let mut n = 0;
+    for id in ids.iter() {
+        let dt = t0.elapsed();
+        if n >= min_blocks && dt > max_duration {
+            info!(
+                "stopped incremental delete after {}us and {} blocks",
+                dt.as_micros(),
+                n
+            );
+            break;
+        }
+        trace!("deleting block for id {}", id);
+        delete_stmt.execute(&[id])?;
+        n += 1;
+    }
+    Ok(n == ids.len())
+}
+
+pub(crate) fn delete_temp_pin(txn: &Transaction, alias: i64) -> rusqlite::Result<()> {
+    txn.prepare_cached("DELETE FROM temp_pins WHERE id = ?")?
+        .execute(&[alias])?;
+    Ok(())
+}
+
+pub(crate) fn put_block<C: ToSql>(
+    txn: &Transaction,
+    key: &C,
+    data: &[u8],
+    links: impl IntoIterator<Item = C>,
+    alias: Option<&AtomicI64>,
+) -> crate::Result<i64> {
+    let id = get_or_create_id(&txn, &key)?;
+    let block_exists = txn
+        .prepare_cached("SELECT 1 FROM blocks WHERE block_id = ?")?
+        .query_row(&[id], |_| Ok(()))
+        .optional()?
+        .is_some();
+    // create a temporary alias for the block, even if it already exists
+    if let Some(alias) = alias {
+        let alias_id = alias.load(Ordering::SeqCst);
+        if alias_id > 0 {
+            txn.prepare_cached("INSERT OR IGNORE INTO temp_pins (id, block_id) VALUES (?, ?)")?
+                .execute(&[alias_id, id])?;
+        } else {
+            // since we are not using an autoincrement column, this will reuse ids.
```

Comment:
transactions are serialized, so no. Using autoincrement won't work since the first column is not unique. It is a m:n mapping, so we reuse the id, possibly quite often.